runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/ArtifactSource.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.artifact;
+
+/**
+ * Makes artifacts available to an ArtifactRetrievalService.
+ */
+public interface ArtifactSource { }
```

Comment:
Can you expand on exactly how and when this is used?

If this is an indirect reference to the [`DistribtedCache`](https://ci.apache.org/projects/flink/flink-docs-release-1.4/api/java/org/apache/flink/api/common/cache/DistributedCache.html), I'm not sure we want to persist and share this between jobs. On the other hand, as far as Flink is concerned, all jobs submitted to the same Job Server effectively run within the context of a single job. At the very least, we need to verify that the same `DistributedCache` can be shared between operator instances within the same job.

Reply:
I was operating under the impression that there would be one Flink job per Beam job, with the Job server creating a new Flink job and context for each submitted job.  If that's not the direction we're going, then I'm doubtful that we should be exposing all keys within the DistributedCache.  If we still want to use it, then we shoud probably wrap it in some validation logic that checks if the artifact an environment is asking for belongs to its assigned job.  Having multiple Beam jobs share the same Flink resources this way will definitely add a layer of complexity.

Reply:
After looking into the how the `DistributedCache` is created in our case (within a `RichMapPartitionFunction` and on a distributed executor), it looks like the DistributedCache is created one and is fixed and reused throughout a given task.

Specifically, the distributed cache entries populated on [Task creation](https://github.com/apache/flink/blob/8395508b0401353ed07375e22882e7581d46ac0e/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L635).

See the following classes for details:
https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java
https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/DistributedRuntimeUDFContext.java
https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java
https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java

Reply:
Based on my previous comment, it looks like we _must_ mint a new `DistributedCache` per submitted Beam job. We want the option of allowing multiple jobs to be submitted to a Job service, so we will need to account for this.

Since we're already munging around with shared cache references, I think it makes the most sense if we require each operator to pass a reference to its own cache when it wants an environment reference. The first call with that job id will provide the backing storage, but multiple job ids need to be supported.